File: unit_tests/test_queen.py
import unittest
from board import Board
from constants import PieceType, Color, Direction


class TestSuite_QueenMoves(unittest.TestCase):
    def test_queen_moves_1(self):
        """Queen's moves on an empty board"""
        for square in ['a1', 'c2', 'e4', 'g7']:
            b1, b2, b3 = Board(), Board(), Board()
            queen = b1.place_piece(square, PieceType.QUEEN, Color.WHITE)
            rook = b2.place_piece(square, PieceType.ROOK, Color.WHITE)
            bishop = b3.place_piece(square, PieceType.BISHOP, Color.WHITE)
            actualMoves = queen.calc_potential_moves()
            expectedMoves = rook.calc_potential_moves() + bishop.calc_potential_moves()
            self.assertEqual(set(map(lambda x: str(x)[1:], actualMoves)), set(map(lambda x: str(x)[1:], expectedMoves)))

    def test_queen_moves_2(self):
        """Queen's moves on a non-empty board"""
        b = Board()
        queen = b.place_piece('g2', PieceType.QUEEN, Color.WHITE)
        b.place_piece('h2', PieceType.KNIGHT, Color.WHITE)
        b.place_piece('f3', PieceType.PAWN, Color.WHITE)
        b.place_piece('h3', PieceType.ROOK, Color.BLACK)
        b.place_piece('g4', PieceType.PAWN, Color.BLACK)
        actualMoves = queen.calc_potential_moves()
        expectedMoves = ['Qg2-a2', 'Qg2-b2', 'Qg2-c2', 'Qg2-d2', 'Qg2-e2', 'Qg2-f2', 'Qg2-f1', 'Qg2-g1', 'Qg2-h1', 'Qg2-g3', 'Qg2-g4', 'Qg2-h3']
        self.assertEqual(set(map(str, actualMoves)), set(expectedMoves))

    def test_queen_moves_pinned(self):
        """Queen's moves while pinned"""
        # there is a queen and a king lined up, but no pinner
        b = Board()
        queen = b.place_piece('c1', PieceType.QUEEN, Color.WHITE)
        b.place_piece('a1', PieceType.KING, Color.WHITE)

        with self.assertRaises(AssertionError):
            queen.calc_potential_moves_pinned(Direction.LEFT)
        with self.assertRaises(AssertionError):
            queen.calc_potential_moves_pinned(Direction.RIGHT)

        # there is a queen and a pinner lined up, but no king
        b = Board()
        queen = b.place_piece('c1', PieceType.QUEEN, Color.WHITE)
        b.place_piece('e1', PieceType.ROOK, Color.BLACK)
        b.place_piece('b1', PieceType.KING, Color.WHITE)
        actualMoves = queen.calc_potential_moves_pinned(Direction.LEFT)
        expectedMoves = ['Qc1-d1', 'Qc1-e1']
        self.assertEqual(set(map(str, actualMoves)), set(expectedMoves))

File: c_engine/test/test_queen.cpp
#include "constants.h"
#include "utest.h"
#include "queen.h"
#include "move.h"

#define private public

#include "board.h"

TestSuite create_test_suite_queen() {
    TestSuite testSuite("Queen");

    testSuite.add_test("Queen construction", []() {
        Board b;
        auto queen = dynamic_cast<Queen*>(b.place_piece(PieceType::QUEEN, Color::WHITE, "a1"));

        assert_equal(PieceType::QUEEN, queen->_kind);
        assert_equal(Color::WHITE, queen->_color);
        assert_equal("a1", queen->_square->get_name());
        assert_equal("Q", queen->_name);
        assert_equal(false, queen->_isLight);

        assert_equal(8U, queen->_slidingDirections.size());
        for (auto dir : {Direction::DOWN, Direction::UP, Direction::RIGHT, Direction::LEFT, Direction::DOWN_RIGHT,
            Direction::DOWN_LEFT, Direction::UP_RIGHT, Direction::UP_LEFT}) {
            assert_vector_contains(queen->_slidingDirections, dir);
            }
    });

    testSuite.add_test("Queen move generation 1", []() {
        Board b;
        auto queen = dynamic_cast<Queen*>(b.place_piece(PieceType::QUEEN, Color::WHITE, "a1"));
        queen->recalculate();

        auto moves = b.squares_to_moves(queen->get_potential_squares(), queen);

        assert_equal(21U, moves.size());
        for (auto move : {"Qa1-a2", "Qa1-a3", "Qa1-a4", "Qa1-a5", "Qa1-a6", "Qa1-a7", "Qa1-a8", "Qa1-b1", "Qa1-c1",
            "Qa1-d1", "Qa1-e1", "Qa1-f1", "Qa1-g1", "Qa1-h1", "Qa1-b2", "Qa1-c3", "Qa1-d4", "Qa1-e5", "Qa1-f6",
            "Qa1-g7", "Qa1-h8"})
            assert_vector_contains(moves, move);
    });

    testSuite.add_test("Queen move generation 2", []() {
        Board b;
        auto queen = dynamic_cast<Queen*>(b.place_piece(PieceType::QUEEN, Color::WHITE, "d4"));
        queen->recalculate();

        auto moves = b.squares_to_moves(queen->get_potential_squares(), queen);

        assert_equal(27U, moves.size());
        for (auto move : {"Qd4-a1", "Qd4-b2", "Qd4-c3", "Qd4-e5", "Qd4-f6", "Qd4-g7", "Qd4-h8", "Qd4-c5", "Qd4-b6",
            "Qd4-a7", "Qd4-e3", "Qd4-f2", "Qd4-g1", "Qd4-d1", "Qd4-d2", "Qd4-d3", "Qd4-d5", "Qd4-d6", "Qd4-d7",
            "Qd4-d8", "Qd4-a4", "Qd4-b4", "Qd4-c4", "Qd4-e4", "Qd4-f4", "Qd4-g4", "Qd4-h4"})
            assert_vector_contains(moves, move);
    });

    testSuite.add_test("Queen move while pinned 1", []() {
        Board b;
        auto queen = dynamic_cast<Queen*>(b.place_piece(PieceType::QUEEN, Color::WHITE, "c2"));
        b.place_piece(PieceType::KING, Color::WHITE, "b1");
        b.place_piece(PieceType::QUEEN, Color::BLACK, "e4");

        queen->recalculate();

        auto moves = b.squares_to_moves(queen->calc_potential_squares_pinned(Direction::UP_RIGHT), queen);

        assert_equal(2U, moves.size());
        assert_vector_contains(moves, "Qc2-d3");
        assert_vector_contains(moves, "Qc2-e4");

        for (auto move : moves)
            delete move;
    });

    return testSuite;
}   